duckdb-vortex/src/expr/expr.cpp

```diff
@@ -0,0 +1,216 @@
+		dtype->mutable_primitive()->set_type(vortex::dtype::F32);
+		return dtype;
+	case LogicalTypeId::DOUBLE:
+		dtype->mutable_primitive()->set_nullable(nullable);
+		dtype->mutable_primitive()->set_type(vortex::dtype::F64);
+		return dtype;
+	case LogicalTypeId::CHAR:
+	case LogicalTypeId::VARCHAR:
+		dtype->mutable_utf8()->set_nullable(nullable);
+		return dtype;
+	case LogicalTypeId::BLOB:
+		dtype->mutable_binary()->set_nullable(nullable);
+		return dtype;
+	default:
+		break;
+	}
+	throw Exception(ExceptionType::NOT_IMPLEMENTED, "into_vortex_dtype", {{"id", type_.ToString()}});
+}
+
+vortex::scalar::Scalar *into_null_scalar(Arena &arena, LogicalType &type_) {
+	auto scalar = Arena::Create<vortex::scalar::Scalar>(&arena);
+	scalar->set_allocated_dtype(into_vortex_dtype(arena, type_, true));
+	scalar->mutable_value()->set_null_value(google::protobuf::NULL_VALUE);
+	return scalar;
+}
+
+vortex::scalar::Scalar *into_vortex_scalar(Arena &arena, Value &value, bool nullable) {
+	auto scalar = Arena::Create<vortex::scalar::Scalar>(&arena);
+	auto dtype = into_vortex_dtype(arena, value.type().id(), nullable);
+	scalar->set_allocated_dtype(dtype);
+
+	switch (value.type().id()) {
+	case LogicalTypeId::INVALID:
+	case LogicalTypeId::SQLNULL: {
+		scalar->mutable_value()->set_null_value(google::protobuf::NULL_VALUE);
+		return scalar;
+	}
+	case LogicalTypeId::BOOLEAN: {
+		auto boolean = new vortex::dtype::Bool();
+		boolean->set_nullable(nullable);
+		scalar->mutable_value()->set_bool_value(value.GetValue<bool>());
+		return scalar;
+	}
+	case LogicalTypeId::TINYINT:
+		scalar->mutable_value()->set_int8_value(value.GetValue<int8_t>());
+		return scalar;
+	case LogicalTypeId::SMALLINT:
+		scalar->mutable_value()->set_int16_value(value.GetValue<int16_t>());
+		return scalar;
+	case LogicalTypeId::INTEGER:
+		scalar->mutable_value()->set_int32_value(value.GetValue<int32_t>());
+		return scalar;
+	case LogicalTypeId::BIGINT:
+		scalar->mutable_value()->set_int64_value(value.GetValue<int64_t>());
+		return scalar;
+	case LogicalTypeId::UTINYINT:
+		scalar->mutable_value()->set_uint8_value(value.GetValue<uint8_t>());
+		return scalar;
+	case LogicalTypeId::USMALLINT:
+		scalar->mutable_value()->set_uint16_value(value.GetValue<uint16_t>());
+		return scalar;
+	case LogicalTypeId::UINTEGER:
+		scalar->mutable_value()->set_uint32_value(value.GetValue<uint32_t>());
+		return scalar;
+	case LogicalTypeId::UBIGINT:
+		scalar->mutable_value()->set_uint64_value(value.GetValue<uint64_t>());
+		return scalar;
+	default:
+		break;
+	}
+	throw Exception(ExceptionType::NOT_IMPLEMENTED, "into_vortex_scalar", {{"id", value.ToString()}});
+}
+
+void set_column(const string &s, vortex::expr::Expr *column) {
+	auto get_item = new vortex::expr::Kind_GetItem();
+	get_item->mutable_path()->assign(s);
+	auto kind = column->mutable_kind();
+	kind->set_allocated_get_item(get_item);
+	column->add_children()->mutable_kind()->set_allocated_identity(new vortex::expr::Kind_Identity());
```

Comment:
Can't find any corresponding `delete` in this PR. Does this currently leak? Can we use a smart pointer instead?

Reply:
fixed